File: R/synthetic_name_counts.R
#' Generate a fake dataset of name counts with “garbled” variants
#'
#' This function builds a synthetic name‐count dataset by starting from the
#' built-in `state.name` vector, replicating each name, applying random
#' character substitutions (“garbling”), and then counting the frequencies of
#' both clean and garbled names.
#'
#' @return A data frame with two columns:
#'   - `name`: character, the (possibly garbled) name
#'   - `count`: integer, the number of times it appears
#'
#' @examples
#' # Generate a toy dataset
#' df <- synthetic_name_counts()
#' head(df)
#'
#' @export
synthetic_name_counts <- function(){

  # Starting with the set of state names
  names = state.name

  # Defining a function to garble names
  garble <- function(x) {
    # split to chars
    s <- strsplit(x, "", TRUE)[[1]]
    L <- length(s)

    # how many substitutions (capped by string length)
    n <- sample(1:3, 1, prob = c(0.6,0.3,0.1))

    # pick n distinct positions and substitute
    pos <- sample.int(L, n)
    s[pos] <- sample(c(letters, rep("",30)), n, replace = TRUE)

    paste(s, collapse = "")
  }

  # Generating the list of clean names
  clean_list <- unlist(lapply(seq_along(names), function(i) {
    rep(names[i], floor(2000 / (i + i)))
  }))

  # Garble each clean name once
  garbled_list <- unlist(lapply(clean_list, garble))

  # Combine clean (10×) and garbled once
  full_list <- c(rep(clean_list, 10), garbled_list)

  # Collapsing to character counts
  name_counts <- full_list |>
    as.data.frame(stringsAsFactors = FALSE) |>
    setNames("name") |>
    dplyr::group_by(name) |>
    dplyr::summarise(count = dplyr::n(), .groups = "drop") |>
    dplyr::arrange(-count) |>
    as.data.frame()

  return(name_counts)

}

File: nicknamer/synthetic_name_counts.py
"""Generate a synthetic dataset of garbled name counts (for testing/demos)."""

import math
import random
import string
from collections import Counter
from typing import List

import pandas as pd


# US state names (equivalent to R's built-in `state.name`)
_STATE_NAMES: List[str] = [
    "Alabama", "Alaska", "Arizona", "Arkansas", "California", "Colorado",
    "Connecticut", "Delaware", "Florida", "Georgia", "Hawaii", "Idaho",
    "Illinois", "Indiana", "Iowa", "Kansas", "Kentucky", "Louisiana",
    "Maine", "Maryland", "Massachusetts", "Michigan", "Minnesota",
    "Mississippi", "Missouri", "Montana", "Nebraska", "Nevada",
    "New Hampshire", "New Jersey", "New Mexico", "New York",
    "North Carolina", "North Dakota", "Ohio", "Oklahoma", "Oregon",
    "Pennsylvania", "Rhode Island", "South Carolina", "South Dakota",
    "Tennessee", "Texas", "Utah", "Vermont", "Virginia", "Washington",
    "West Virginia", "Wisconsin", "Wyoming",
]


def _garble(x: str) -> str:
    """Apply 1–3 random character substitutions (or deletions) to *x*."""
    s = list(x)
    L = len(s)
    n = random.choices([1, 2, 3], weights=[0.6, 0.3, 0.1])[0]
    n = min(n, L)
    positions = random.sample(range(L), n)
    # Pool: all lower-case letters + 30 empty strings (deletions)
    pool = list(string.ascii_lowercase) + [""] * 30
    for pos in positions:
        s[pos] = random.choice(pool)
    return "".join(s)


def synthetic_name_counts() -> pd.DataFrame:
    """Generate a fake dataset of name counts with random garbled variants.

    Starts from the 50 US state names, replicates each name (count depends on
    rank), applies random character substitutions ("garbling"), then tallies
    both clean and garbled occurrences.

    Returns
    -------
    pandas.DataFrame with columns:
        ``name``  – the (possibly garbled) name.
        ``count`` – the number of times it appears.
    """
    names = _STATE_NAMES

    # Build clean list: name[i] repeated floor(2000 / (2*(i+1))) times
    clean_list: List[str] = []
    for i, name in enumerate(names):
        reps = math.floor(2000 / (2 * (i + 1)))
        clean_list.extend([name] * reps)

    # Garble each clean name once
    garbled_list = [_garble(n) for n in clean_list]

    # Combine: 10× clean + 1× garbled
    full_list = clean_list * 10 + garbled_list

    counts = Counter(full_list)
    df = (
        pd.DataFrame(counts.items(), columns=["name", "count"])
        .sort_values("count", ascending=False)
        .reset_index(drop=True)
    )
    return df
